service/glacier/internal/customizations/treehash.go

```diff
@@ -0,0 +1,144 @@
+package customizations
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"github.com/awslabs/smithy-go"
+	"github.com/awslabs/smithy-go/middleware"
+	smithyhttp "github.com/awslabs/smithy-go/transport/http"
+	"io"
+)
+
+const bufsize = 1024 * 1024
+
+// AddTreeHashMiddleware adds middleware needed to automatically
+// calculate Glacier's required checksum headers.
+func AddTreeHashMiddleware(stack *middleware.Stack) {
+	stack.Finalize.Add(&TreeHashMiddleware{}, middleware.Before)
+}
+
+// TreeHashMiddleware provides the middleware that will automatically
+// set the sha256 and tree hash headers if they have not already been
+// set.
+type TreeHashMiddleware struct{}
+
+// ID returns the middleware ID.
+func (*TreeHashMiddleware) ID() string { return "Glacier:TreeHash" }
+
+// HandleFinalize implements the finalize middleware handler method
+func (*TreeHashMiddleware) HandleFinalize(
+	ctx context.Context, input middleware.FinalizeInput, next middleware.FinalizeHandler,
+) (
+	output middleware.FinalizeOutput, metadata middleware.Metadata, err error,
+) {
+	req, ok := input.Request.(*smithyhttp.Request)
+	if !ok {
+		return output, metadata, &smithy.SerializationError{
+			Err: fmt.Errorf("unknown request type %T", input.Request),
+		}
+	}
+
+	if err := addChecksum(req); err != nil {
+		return output, metadata, &smithy.SerializationError{Err: err}
+	}
+
+	return next.HandleFinalize(ctx, input)
+}
+
+func addChecksum(req *smithyhttp.Request) error {
+	if req.GetStream() == nil || req.Header.Get("X-Amz-Sha256-Tree-Hash") != "" {
+		return nil
+	}
+
+	h := computeHashes(req.GetStream())
+	if err := req.RewindStream(); err != nil {
+		return err
+	}
+
+	hstr := hex.EncodeToString(h.TreeHash)
+	req.Header.Set("X-Amz-Sha256-Tree-Hash", hstr)
+
+	hLstr := hex.EncodeToString(h.LinearHash)
+	req.Header.Set("X-Amz-Content-Sha256", hLstr)
+
+	return nil
+}
+
+// Hash contains information about the tree-hash and linear hash of a
+// Glacier payload. This structure is generated by computeHashes().
+type Hash struct {
+	TreeHash   []byte
+	LinearHash []byte
+}
+
+// ComputeHashes computes the tree-hash and linear hash of a reader r.
+//
+// See http://docs.aws.amazon.com/amazonglacier/latest/dev/checksum-calculations.html for more information.
+func ComputeHashes(r io.ReadSeeker) Hash {
+	start, _ := r.Seek(0, io.SeekCurrent) // Read the whole stream
+	defer r.Seek(start, io.SeekStart)     		// Rewind stream at end
+	return computeHashes(r)
+}
```

Comment:
`ComputeHashes` and `computeHashes` is a little confusing since it looks like only the lower case `computeHashes` is used.

Reply:
Upper case is what's exposed for customers to use. These functions are exposed in the existing version of V2 so I wanted to keep that. Not exposing them would definitely be simpler if that's where we want to go.

Reply:
In this case this func is nested under an `internal` package so won't be visible to them.